Documentation/rfc/v3api.md

```diff
@@ -0,0 +1,459 @@
+
+
+## Protobuf Defined API
+
+``` protobuf
+syntax = "proto3";
+
+// Interface exported by the server.
+service etcd {
+  // Range gets the keys in the range from the store.
+  // Maybe rename it to GET?
+  rpc Range(RangeRequest) returns (RangeResponse) {}
+
+  // Put puts the given key into the store.
+  // A put request increases the index of the store,
+  // and generates one event in the event history.
+  rpc Put(PutRequest) returns (PutResponse) {}
+
+  // Delete deletes the given key from the store.
+  // A delete request increase the index of the store,
+  // and generates one event in the event history.
+  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+
+  // Tnx processes all the requests in one transaction.
+  // A tnx request increases the index of the store,
+  // and generates one event in the event history.
+  rpc Tnx(TnxRequest) returns (TnxResponse) {}
+
+  // Watch watches the events happening or happened in etcd. Both input and output
+  // are stream. One watch rpc can watch for multiple ranges and get a stream of
+  // events. The whole events history can be watched unless compacted.
+  rpc Watch(stream WatchRequest) returns (stream WatchResponse) {}
+
+  // Compact compacts the event history in etcd. User should compact the
+  // event history periodically, or it will grow infinitely.
+  rpc Compact(CompactRequest) returns (CompactResponse) {}
+
+  // LeaseCreate creates a lease. A lease has a TTL. The lease will expire if the
+  // server does not receive a keepAlive within TTL from the lease holder.
+  // All keys attached to the lease will be expired and deleted if the lease expires.
+  // The key expiration generates an event in event history.
+  rpc LeaseCreate(LeaseCreateRequest) returns (LeaseCreateResponse) {}
+
+  // LeaseRevoke revokes a lease. All the key attached to the lease will be expired and deleted.
+  rpc LeaseRevoke(LeaseRevokeRequest) returns (LeaseRevokeResponse) {}
+
+  // LeaseAttach attaches keys with a lease.
+  rpc LeaseAttach(LeaseAttachRequest) returns (LeaseAttachResponse) {}
+
+  // LeaseTnx likes Tnx. It does one additional thing that all the keys put into the
+  // store are attached with the given lease. This is rpc is useful when you want to
+  // put a key and attach it to a lease atomically.
+  rpc LeaseTnx(LeaseTnxRequest) returns (LeaseTnxResponse) {}
+
+  // KeepAlive keeps the lease alive.
+  rpc LeaseKeepAlive(stream LeaseKeepAliveRequest) returns (stream LeaseKeepAliveResponse) {}
+}
+
+message RequestHeader {
+  // the cluster id to check with
+  // if the cluster id is not matched, an error is returned.
+  optional uint64 cluster_id = 1;
+}
+
+message ResponseHeader {
+  // an error type message?
+  optional string error = 1;
+  optional uint64 cluster_id = 2;
+  optional uint64 member_id = 3;
+  // index of the store when the requested was processed.
+  optional int64 index = 4;
+  optional uint64 raft_term = 5;
+}
+
+message RangeRequest {
```

Comment:
> On Apr 14, 2015, at 12:03 PM, Xiang Li <notifications@github.com> wrote:
> 
> In Documentation/rfc/v3api.md:
> 
> > +  // the cluster id to check with
> > +  // if the cluster id is not matched, an error is returned.
> > +  optional uint64 cluster_id = 1;
> > +}
> > +
> > +message ResponseHeader {
> > +  // an error type message?
> > +  optional string error = 1;
> > +  optional uint64 cluster_id = 2;
> > +  optional uint64 member_id = 3;
> > +  // index of the store when the requested was processed.
> > +  optional int64 index = 4;
> > +  optional uint64 raft_term = 5;
> > +}
> > +
> > +message RangeRequest {
> @smarterclayton
> 
> The current API does not directly support consistent range over multiple requests. You can build your own via range + history.
> 
> Do you have a real use case? What are you trying to solve?
> 
For Kubernetes, there are advantages to being able to do ranged snapshot queries to process data in chunks (instead of loading 100k keys in memory at once, make 100 1K range requests).  We can do that via buckets in the keyspace, but the larger appeal of range was the consistent chunk aspect.

That being said, it does not break our use cases if that's not available.  I believe in order to emulate that we would need delete tombstones in the key space and would have to track the highest modified index in the scan.  Do you anticipate/expect there to be delete tombstones in the keyspace?
> —
> Reply to this email directly or view it on GitHub.
>